hcl2template/addrs/plugin.go

```diff
@@ -0,0 +1,238 @@
+//
+// The following are valid source string formats:
+// 		name
+// 		namespace/name
+// 		hostname/namespace/name
+func ParsePluginSourceString(str string) (*Plugin, hcl.Diagnostics) {
+	ret := &Plugin{
+		Hostname:  DefaultPluginHost,
+		Namespace: DefaultPluginNamespace,
+	}
+	var diags hcl.Diagnostics
+
+	// split the source string into individual components
+	parts := strings.Split(str, "/")
+	if len(parts) == 0 || len(parts) > 3 {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid plugin source string",
+			Detail:   `The "source" attribute must be in the format "[hostname/][namespace/]name"`,
+		})
+		return nil, diags
+	}
+
+	// check for an invalid empty string in any part
+	for i := range parts {
+		if parts[i] == "" {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid plugin source string",
+				Detail:   `The "source" attribute must be in the format "[hostname/][namespace/]name"`,
+			})
+			return nil, diags
+		}
+	}
+
+	// check the 'name' portion, which is always the last part
+	givenName := parts[len(parts)-1]
+	name, err := ParsePluginPart(givenName)
+	if err != nil {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid plugin type",
+			Detail:   fmt.Sprintf(`Invalid plugin type %q in source %q: %s"`, givenName, str, err),
+		})
+		return nil, diags
+	}
+	ret.Type = name
+
+	if len(parts) == 1 {
+		return ret, diags
+	}
+
+	if len(parts) >= 2 {
+		// the namespace is always the second-to-last part
+		givenNamespace := parts[len(parts)-2]
+		namespace, err := ParsePluginPart(givenNamespace)
+		if err != nil {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid plugin namespace",
+				Detail:   fmt.Sprintf(`Invalid plugin namespace %q in source %q: %s"`, namespace, str, err),
+			})
+			return nil, diags
+		}
+		ret.Namespace = namespace
+	}
+
+	// Final Case: 3 parts
+	if len(parts) == 3 {
+		// the hostname is always the first part in a three-part source string
+		hostname := parts[0]
+		// TODO(azr): validate host ? Can this be something else than a
+		// github.com host for now?
+		ret.Hostname = hostname
+	}
+
+	// Due to how plugin executables are named and plugin git repositories
+	// are conventionally named, it's a reasonable and
+	// apparently-somewhat-common user error to incorrectly use the
+	// "packer-plugin-" prefix in a plugin source address. There is
+	// no good reason for a plugin to have the prefix "packer-" anyway,
+	// so we've made that invalid from the start both so we can give feedback
+	// to plugin developers about the packer- prefix being redundant
+	// and give specialized feedback to folks who incorrectly use the full
+	// packer-plugin- prefix to help them self-correct.
+	const redundantPrefix = "packer-"
+	const userErrorPrefix = "packer-plugin-"
+	if strings.HasPrefix(ret.Type, redundantPrefix) {
```

Comment:
I'm also struggling here. Let's use a repo that used to be pretty common: 

https://github.com/jetbrains-infra/packer-builder-vsphere

The repo is named that because it's also what it should be named on install.

I _think_ this code is to tell people they needn't name their repos things like packer-plugin-party but I don't see how this is bad, and even if it is why this is such a bad thing we need to enforce against it.

Reply:
This pkg is mainly taken out from Terraform and this error case made sense to me so I left it.

It's in the case someone wants to require a packer plugin and names it like so:

```hcl
packer {
    required_plugins {
        packer-plugin-amazon = ">= v0" // err: should be "amazon" instead of packer-plugin-amazon
        amazon-v1 = {
            source  = "packer-plugin-amazon" // should be "amazon"
            version = ">= v1"
        }
       // ...
   }
}
```

I think this is really nice for consistency across packer configs and with Terraform.

Reply:
aha, so the name being called in the _template_ should not contain "packer-plugin" but we're not nitpicking the name of the repo or whatever. 

Let's make sure to link this error to a docs page with an explanation like you gave above, because that makes it much clearer.